### PR TITLE
Implemented a slue of new functions to interact with the realms API

### PR DESCRIPTION
**Pull request by IanTapply22**
This PR implements a handful of new functions to interact with even more parts of the Minecraft realms API—all the way from banning players to deleting a realm with the specified ID. You can find the added functionality below:

## Bedrock & Java 
- Get realm subscription info
- Get detailed realm subscription info
- Change active realm slot
- Change realm name and description
- Delete realm

## Bedrock
- Reset realm
- Update realm configuration
- Remove realm invite
- OP realm player
- DEOP realm player
- Ban player from the realm
- Unban player from the realm
- Remove realm from joined list
- Update forced texture pack rule
- Update realm default permission
- Update realm player permission

Feel free to leave suggestions or comments if you need anything fixed, added, or updated. Thanks

### CODE REVIEW

**Comment by IanTapply22** [comment]
Are we wanting to add tests for all of these calls?

File: src/bedrock/api.js [hunk 1]
```diff
@@ -94,8 +94,122 @@ module.exports = class BedrockRealmAPI extends RealmAPI {
...
+
+  async unbanPlayerFromRealm (realmId, uuid) {
+    return await this.rest.delete(`/worlds/${realmId}/blocklist/${uuid}`)
+  }
+
+  async removeRealmFromJoinedList (realmId) {
+    return await this.rest.delete(`/invites/${realmId}`)
+  }
+
+  async changeForcedTexturePack (realmId, forced) {
```
Comment (extremeheat) [inline src/bedrock/api.js:190]: changeForcedTexturePack here sounds clearer to me as "changeIsTexturePackRequired"

**Review by extremeheat — CHANGES_REQUESTED** [review]
Thanks for working on this. Left some comments, but yes tests would be good for API calls that are expected to return something from the server, so we know what data we're working with.

These new functions would also need to be added to the doc, with explicit information on what they return.

cc @LucienHH if you have anything to add

**Comment by IanTapply22** [comment]
Sounds good! Thank you for the feedback and the review. Expect everything to be pushed by the end of today. (in ~4 hours from now max)

File: src/structures/Realm.js [hunk 2]
```diff
@@ -45,11 +45,31 @@ module.exports = class Realm {
...
+  async getSubscriptionInfoDetailed () {
+    return this.#api.getRealmSubscriptionInfoDetailed(this.id)
+  }
+
+  async changeActiveSlot () {
+    return this.#api.changeRealmActiveSlot(this.id, this.activeSlot)
+  }
+
+  async changeNameAndDescription () {
+    return this.#api.changeRealmNameAndDescription(this.id, this.name, this.motd)
```
[the remark this answers is not in the document]
Comment (IanTapply22) [reply src/structures/Realm.js:73]: If the realmID is the only parameter for a function, would I just leave the parameters blank to follow with the other functions in this file?

File: src/bedrock/api.js [hunk 3]
```diff
@@ -94,8 +94,122 @@ module.exports = class BedrockRealmAPI extends RealmAPI {
...
+    })
+  }
+
+  async removeAllRealmPlayers (realmId) {
+    const data = await this.rest.put(`/invites/${realmId}/invite/update`, {
+      body: {
+        invites: null
+      }
+    })
+    return new Realm(this, data)
```
[the remark this answers is not in the document]
Comment (IanTapply22) [reply src/bedrock/api.js:175]: It actually seems that all invite endpoints are supposed to return a realm object. Especially the removeAllRealmPlayers function

[the remark this answers is not in the document]
Comment (IanTapply22) [reply src/bedrock/api.js:175]: Nevermind, I stand corrected. I don't need to return a new Realm for that. I can just return the result from the request. Sorry

File: src/structures/Realm.js (hunk 2 above)
[the remark this answers is not in the document]
Comment (extremeheat) [reply src/structures/Realm.js:73]: No, I mean this code doesn't make any sense. How can you "changeNameAndDescription" without passing any parameters ?

**Comment by IanTapply22** [comment]
For now, I'm not pushing anything related to removing all realm players for now. There seems to be an issue with setting values as null in the body of the requests.

I will be creating another PR for this issue to address it better and to keep it separated from here.

File: src/structures/Realm.js (hunk 2 above)
[the remark this answers is not in the document]
Comment (IanTapply22) [reply src/structures/Realm.js]: Fixed

File: src/bedrock/api.js (hunk 3 above)
[the remark this answers is not in the document]
Comment (IanTapply22) [reply src/bedrock/api.js]: Going to be fixed in separate PR to address an issue with null vlaues in body of requests

File: src/bedrock/api.js [hunk 4]
```diff
@@ -94,8 +94,113 @@ module.exports = class BedrockRealmAPI extends RealmAPI {
...
+  }
+
+  // Reference https://github.com/PrismarineJS/prismarine-realms/issues/34 for configuration structure
+  async changeRealmConfiguration (realmId, configuration) {
+    return await this.rest.put(`/worlds/${realmId}/configuration`, {
+      body: configuration
+    })
+  }
+
+  async removeRealmInvite (realmId, uuid) {
```
Comment (LucienHH) [inline src/bedrock/api.js:138]: Maybe `removePlayerFromRealm` may be a better name here to draw a clear distinction from `rejectRealmInvite`.

File: src/bedrock/api.js [hunk 5]
```diff
@@ -94,8 +94,113 @@ module.exports = class BedrockRealmAPI extends RealmAPI {
...
+      }
+    })
+    return new Realm(this, data)
+  }
+
+  async changeRealmPlayerPermission (realmId, permission, uuid) {
+    return await this.rest.put(`/world/${realmId}/userPermission`, {
+      body: {
+        permission: permission.toUpperCase(),
+        uuid
```
Comment (LucienHH) [inline src/bedrock/api.js:202]: This endpoint on Bedrock expects `{ xuid: string, permission: string }`

File: src/java/api.js [hunk 6]
```diff
@@ -24,8 +24,33 @@ module.exports = class JavaRealmAPI extends RealmAPI {
...
+  async deleteRealm (realmId) {
+    return await this.rest.delete(`/worlds/${realmId}`)
+  }
+
   async getRealmWorldDownload (realmId, slotId) {
     const data = await this.rest.get(`/worlds/${realmId}/slot/${slotId}/download`)
     return new Download(this, data)
   }
+
+  async getRealmSubscriptionInfo (realmId) {
```
Comment (LucienHH) [inline src/java/api.js:36]: If endpoints match on both Java and Bedrock the method to call that endpoint can be declared once in the `RealmAPI` class

File: docs/API.md [hunk 7]
```diff
@@ -301,8 +324,222 @@ Rejects a pending invite for the authenticating account
...
+
+No output
+
+</details>
+
+---
+
+### removeRealmInvite
+
+(realmId: string, uuid: string) => Promise\<Realm>
```
Comment (LucienHH) [inline docs/API.md:338]: Might be useful to note that for Bedrock the player's XUID should be provided as the UUID

**Comment by IanTapply22** [comment]
Fixed everything there, there are a couple of small things that still need to be fixed but I already have another PR cooking up to address them.

File: docs/API.md [hunk 8]
```diff
@@ -471,6 +722,100 @@ await realm.getWorldDownload()
...
+---
+
+#### getSubscriptionInfoDetailed
+
+() => Promise\<RealmSubscriptionInfoDetailed>
+
+Gets the detailed subscription info
+
+```js
+await realm.getSubscriptionInfoDetailed()
```
Comment (extremeheat) [inline docs/API.md:757]: Why do we need two functions getSubscriptionInfo for this? Why not just an extra param to add more data?

**Comment by IanTapply22** [comment]
If you could just make one big list of changes that need to be made it'd be greatly appreciated. It's nice to just have one big push of fixes then the commit history doesn't get all clogged unless we're doing a rebase when merging.

**Comment by extremeheat** [comment]
Don't worry about the PR commits. It will be squashed on merge.

File: docs/API.md [hunk 9]
```diff
@@ -301,8 +324,223 @@ Rejects a pending invite for the authenticating account
...
...ue,texturePacksRequired:true,timeRequest:null,enabledPacks:{resourcePacks:[""],behaviorPacks:[""]},customGameServerGlobalProperties:null,worldSettings:{sendcommandfeedback:{type:0,value:true}commandblocksenabled:{type:0,value:true},dodaylightcycle:{type:0,value:true},randomtickspeed:{type:1,value:3},naturalregeneration:{type:0,value:true},showtags:{type:0,value:true},commandblockoutput:{type:0,value:true},dofiretick:{type:0,value:false},maxcommandchainlength:{type:1,value:65535},falldamage:{type:0,value:true},tntexplodes:{type:0,value:true},drowningdamage:{type:0,value:true},domobloot:{type:0,value:true},domobspawning:{type:0,value:true},showbordereffect:{type:0,value,:true},showdeathmessages:{type:0,value:true},respawnblocksexplode:{type:0,value:true},doweathercycle:{type:0,value:true},doentitydrops:{type:0,value:true},doimmediaterespawn:{type:0,value:true},freezedamage:{type:0,value:true},pvp:{type:0,value:true},keepinventory:{type:0,value:false},doinsomnia:{type:0,value:true},mobgriefing:{type:0,value:true},dotiledrops:{type:0,value:true},firedamage:{type:0,value:true},functioncommandlimit:{type:1,value:10000},spawnradius:{type:1,value:25},showcoordinates:{type:0,value:true}}}}}
```
Comment (LucienHH) [inline docs/API.md:381]: Maybe out of scope for this PR but providing some sort of config builder would make it a lot easier to interact with the API here. I.e. 

```js
const { RealmConfigBuilder } = require('prismarine-realms')

const config = new RealmConfigBuilder()
  .setName('')
  .setDescription('')
  .addWorldSettings(
    { name: 'dofiretick', value: true },
    { name: 'domobloot', value: false}
  )
  
  api.changeRealmConfiguration(config)
  ```

Comment (IanTapply22) [reply docs/API.md:381]: Yep! I was thinking the same thing. It was really bugging me with the large JSON code block. I'll open another PR relatively soon sometime this week, or I'll just push it along with the next PR I have as a draft right now.

**Comment by IanTapply22** [comment]
If you have any feedback for anything else that is out of scope or that you think doesn't belong in this PR just let me know and I'll create another PR. Right now I'm just trying to get a good baseline to start from to expand further and improve.

**Review by extremeheat — COMMENTED** [review]
The purpose of prismarine realms isn't to be an basic API wrapper per se but provide a stable high level lib. If we have to put out a breaking change if the backend changes its data format then it's no good.

So if we either return, store or expose data, it needs to be  explicit. If something is supposedly void then generally there should not be any return value in the method.

**Comment by IanTapply22** [comment]
So I should be making custom types for responses for every function that returns something other than void in index.ts?

**Comment by extremeheat** [comment]
Yes, either don't return anything or doc what is returned

**Comment by IanTapply22** [comment]
Any other changes that need to be made?

File: docs/API.md [hunk 10]
```diff
@@ -116,15 +135,15 @@ await api.getRealmBackups('1234567', '1')
...
+(realmId: string, slotId: string, backupId: string) => Promise\<string>
 
 Restores a Realm from a backup
 
 ```js
 await api.restoreRealmFromBackup('1234567', '1', '1970-01-01T00:00:00.000Z')
 ```
 
-No output
+Either 'Retry again later' or 'true'. Always seems to return 'Retry again later' on the first try.
```
Comment (extremeheat) [inline docs/API.md:146]: Shouldn't this be handled internally with an exception? This doesn't seem like good exposed API.

File: docs/API.md [hunk 11]
```diff
@@ -301,8 +341,223 @@ Rejects a pending invite for the authenticating account
...
+
+No output
+
+</details>
+
+---
+
+### removePlayerFromRealm
+
+(realmId: string, xuid: string) => Promise\<Realm>
```
Comment (extremeheat) [inline docs/API.md:355]: Why specifically return a Realm here?

File: docs/API.md (hunk 9 above)
Comment (extremeheat) [reply docs/API.md:398]: You can remove this function from this PR and add it to a later one if it's not ready

File: docs/API.md [hunk 12]
```diff
@@ -301,8 +341,223 @@ Rejects a pending invite for the authenticating account
...
+
+No output
+
+</details>
+
+---
+
+### opRealmPlayer
+
+(realmId: string, uuid: string) => Promise\<Realm>
```
Comment (extremeheat) [inline docs/API.md:413]: Why return Realm here?

File: docs/API.md [hunk 13]
```diff
@@ -301,8 +341,223 @@ Rejects a pending invite for the authenticating account
...
+
+[Realm](#realm)
+
+</details>
+
+---
+
+### deopRealmPlayer
+
+(realmId: string, uuid: string) => Promise\<Realm>
```
Comment (extremeheat) [inline docs/API.md:432]: Same comment as above. This does but banPlayerFromRealm doesn't return Realm.

File: docs/API.md [hunk 14]
```diff
@@ -407,28 +662,42 @@ await realm.invitePlayer('a8005260a332457097a50bdbe48a9a21', 'Steve')
...
+() => Promise\<boolean>
 
 Opens a Realm. Allows players to join
 
 ```js
 await realm.open()
 ```
 
-No output
+True if the world has opened
```
Comment (extremeheat) [inline docs/API.md:673]: Can't this be an exception? Otherwise example code needs to handle the false case.

**Review by extremeheat — COMMENTED** [review]
Left some comments about API. Some of these things should be exceptions. Otherwise if a return value is expected, like status of success, then the example code needs to handle the false cases when calling functions like .open().

File: docs/API.md (hunk 11 above)
Comment (IanTapply22) [reply docs/API.md:355]: In the bedrock Realm API docs, it states it returns all information about the Realm. Should we not be returning all the Realm's information as states in the API docs?

File: docs/API.md (hunk 12 above)
Comment (IanTapply22) [reply docs/API.md:413]: Same question as [here](https://github.com/PrismarineJS/prismarine-realms/pull/40#discussion_r1372319152)

File: docs/API.md (hunk 14 above)
Comment (IanTapply22) [reply docs/API.md:673]: There is no false case, as stated [here](https://wiki.vg/Bedrock_Realms#PUT_.2Fworlds.2F.7Bid.7D.2Fopen) it doesn't return false, it returns the retry again later string. Should we be returning false from this function considering it's basically false if it says retry again later.

File: docs/API.md (hunk 11 above)
Comment (IanTapply22) [reply docs/API.md:355]: Nevermind, ignore that. I understand now. Thank you

File: docs/API.md (hunk 14 above)
Comment (IanTapply22) [reply docs/API.md:673]: I also understand this too, sorry about that.

**Comment by extremeheat** [comment]
So after a second review, some of the doc/types you updated aren't things changed in the code, so instead of making changes there per my previous comments (about how we should throw), what you can do is simply revert those changes (like changes to .open()). We can handle those later, so we can get this merged quicker.

Also, @LucienHH do you think the configuration object will change or break in the future? If so we should move that out of the PR until some abstraction is implemented there. If not then I think it's ok if we just leave it as an dynamic object.

**Comment by LucienHH** [comment]
> Also, @LucienHH do you think the configuration object will change or break in the future? If so we should move that out of the PR until some abstraction is implemented there. If not then I think it's ok if we just leave it as an dynamic object.

The config can change from update to update so should definitely be abstracted away, funnily enough the config has already changed from the one in this PR.

**Comment by rom1504** [comment]
What is the status of this PR? what's missing to finish it ?

**Comment by rom1504** [comment]
@LucienHH @extremeheat what do you think on this PR?

**Review by extremeheat — APPROVED** [review]
LGTM, I think previous my comments about API design are not worth blocking otherwise good PR. There are some useful additions here, thanks @IanTapply22 for the work on this

**Comment by extremeheat** [comment]
I commented out the `changeRealmConfiguration` method since that's not stable API, we can handle this in another PR

**Comment by extremeheat** [comment]
Thanks for the work on this. We can discuss the remaining changes in the other PR. I'll merge so we can resolve the PR backlog.

**Pull request merged**